[GTK4] Toplevel window state change is not handled properly

carlosgcampos opened the pull request:
#### 989de9dbc7acaa3042ca42a0cc2dabbe8478121a
<pre>
[GTK4] Toplevel window state change is not handled properly
<a href="https://bugs.webkit.org/show_bug.cgi?id=260352">https://bugs.webkit.org/show_bug.cgi?id=260352</a>

Reviewed by Michael Catanzaro.

We monitor the toplevel window state to detect when the window is
minimized to set the web pages as not visible and pause the rendering.
In GTK3 we already get a change mask, so we do nothing if iconized state
hasn&apos;t changed. In GTK4 we get the new state, but we don&apos;t know what
changed, so we are actually changing the visibility of pages for any
state change. We need to handle the changed mask manually.
This patch also improves the way we handle visibility changes in
general, because in GTK4 we get a size allocate after the map, which
means we send the visibility change message before the resize, so end up
resuming rendering with 0x0 size in the page. To try to avoid this we
check if we have a size already on map, before the visibility check, and
set the size to make sure the resize is sent before.

* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
(webkitWebViewBaseSetSize):
(webkitWebViewBaseSizeAllocate):
(webkitWebViewBaseMap):
(toplevelWindowStateChanged):
(toplevelWindowRealized):
(webkitWebViewBaseRoot):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/DrawingAreaCoordinatedGraphics.cpp:
(WebKit::DrawingAreaCoordinatedGraphics::DrawingAreaCoordinatedGraphics):

Canonical link: <a href="https://commits.webkit.org/267032@main">https://commits.webkit.org/267032@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5ec2bd06bd667ac22ebb5177c0edae1ac730ccc6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/15444 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15748 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/16111 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/17198 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14478 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/18264 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15844 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/17063 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15628 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/16042 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/13129 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17939 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/13316 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/20862 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/14392 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/14096 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/17359 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14674 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/12430 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13933 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/18293 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1880 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14495 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro reviewed (approved):
There is a typo in the commit message: "tolevel" -> "toplevel"

carlosgcampos committed:
[GTK4] Toplevel window state change is not handled properly
https://bugs.webkit.org/show_bug.cgi?id=260352

Reviewed by Michael Catanzaro.

We monitor the toplevel window state to detect when the window is
minimized to set the web pages as not visible and pause the rendering.
In GTK3 we already get a change mask, so we do nothing if iconized state
hasn't changed. In GTK4 we get the new state, but we don't know what
changed, so we are actually changing the visibility of pages for any
state change. We need to handle the changed mask manually.
This patch also improves the way we handle visibility changes in
general, because in GTK4 we get a size allocate after the map, which
means we send the visibility change message before the resize, so end up
resuming rendering with 0x0 size in the page. To try to avoid this we
check if we have a size already on map, before the visibility check, and
set the size to make sure the resize is sent before.

* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
(webkitWebViewBaseSetSize):
(webkitWebViewBaseSizeAllocate):
(webkitWebViewBaseMap):
(toplevelWindowStateChanged):
(toplevelWindowRealized):
(webkitWebViewBaseRoot):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/DrawingAreaCoordinatedGraphics.cpp:
(WebKit::DrawingAreaCoordinatedGraphics::DrawingAreaCoordinatedGraphics):

Canonical link: https://commits.webkit.org/267032@main
